src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp

```diff
@@ -539,6 +539,67 @@ void ParallelScavengeHeap::object_iterate(ObjectClosure* cl) {
   old_gen()->object_iterate(cl);
 }
 
+// The HeapBlockClaimer is used during parallel iteration over the heap,
+// allowing workers to claim heap blocks, gaining exclusive rights to these blocks.
+// The eden and survivor spaces are treated as single blocks as it is hard to divide
+// these spaces.
+// The old spaces are divided into serveral fixed-size blocks.
+class HeapBlockClaimer : public StackObj {
+  // Index of iterable block, negative values for indexes of young generation spaces,
+  // zero and positive values for indexes of blocks in old generation space.
+  ssize_t _claimed_index;
+ public:
+  static const ssize_t EdenIndex = -2;
+  static const ssize_t SurvivorIndex = -1;
+
+  HeapBlockClaimer() : _claimed_index(EdenIndex) { }
+  // Claim the block and get the block index.
+  bool claim_and_get_block(ssize_t* block_index) {
+    assert(block_index != NULL, "Invalid index pointer");
+    *block_index = Atomic::fetch_and_add(&_claimed_index, 1);
+    ssize_t iterable_blocks = (ssize_t)ParallelScavengeHeap::heap()->old_gen()->iterable_blocks();
+    if (*block_index >= iterable_blocks) {
+      return false;
+    }
+    return true;
+  }
```

Comment:
I wonder if the interface would be cleaner with returning `block_index` directly, i.e. `ssize_t claim_and_get_block()`.  When the index goes out of range, return an `InvalidIndex`, which could be defined as `-3`, like `EdenIndex` above.

Reply:
Thanks for  your  suggestion. I have thought about it when writing this method, but I was not able to find a value of InvalidIndex. After reconsider it, plus that I will make block_index to be unsigned to avoid signed/unsigned conversion, I think might be the InvalidIndex could be max_size_t (that is (size_t)-1). what do you think?

Reply:
I think the new version looks very nice; thank you for the change.